c++/triqs/atom_diag/impl/atom_diag.cpp

```diff
@@ -107,17 +108,48 @@ namespace triqs {
     auto atom_diag<Complex>::get_matrix_element_of_monomial(operators::monomial_t const &op_vec, int B) const
       -> std::pair<int, matrix_t> {
 
-      auto m = nda::eye<scalar_t>(get_subspace_dim(B));
+      imperative_operator<class hilbert_space, scalar_t, false>
+        monomial_op(many_body_op_t(1.0, op_vec), fops);
 
-      auto const &fops = get_fops();
-      for (int i = op_vec.size() - 1; i >= 0; --i) {
-        int ind = fops[op_vec[i].indices];
-        int Bp  = (op_vec[i].dagger ? cdag_connection(ind, B) : c_connection(ind, B));
-        if (Bp == -1) return {-1, std::move(m)};
-        m = (op_vec[i].dagger ? cdag_matrix(ind, B) : c_matrix(ind, B)) * m;
-        B = Bp;
+      state<class hilbert_space, scalar_t, true> initial_st(full_hs);
+      state<class hilbert_space, scalar_t, true> final_st(full_hs);
+
+      matrix_t m;
+      int Bp = -1;
+      bool found_nonzero_element = false;
+
+      auto const& B_fock_states = get_fock_states(B);
+      for(fock_state_t i_index = 0; i_index < B_fock_states.size(); ++i_index) {
```

Comment:
`i_index` is not really a `fock_state_t` but rather the index of the fock_state?
You could do
```cpp
for(auto [fs_idx, fs]: itertools::enumerate(B_fock_states)) {
  ...
}